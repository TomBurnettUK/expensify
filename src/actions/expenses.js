import database from '../firebase/database';

// 'ADD_EXPENSE'
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => async dispatch => {
  const defaultData = {
    description: '',
    note: '',
    amount: 0,
    createdAt: 0
  };
  const expense = { ...defaultData, ...expenseData };
  const ref = await database.ref('expenses').push(expense);
  dispatch(addExpense({ id: ref.key, ...expense }));
};

// 'REMOVE_EXPENSE'
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => async dispatch => {
  await database.ref(`expenses/${id}`).remove();
  dispatch(removeExpense({ id }));
};

// 'EDIT_EXPENSE'
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// 'SET_EXPENSES'
export const setExpenses = expenses => ({ type: 'SET_EXPENSES', expenses });

export const startSetExpenses = () => async dispatch => {
  const snapshot = await database.ref('expenses').once('value');
  const expenses = [];
  snapshot.forEach(expense => {
    expenses.push({ ...expense.val(), id: expense.key });
  });
  dispatch(setExpenses(expenses));
};
