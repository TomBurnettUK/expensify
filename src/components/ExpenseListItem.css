.list-item {
  border: 1px solid #D2D2D2;
  border-top: none;
  color: #333;
  padding: 0 1rem;
  text-decoration: none;
  display: flex;
  flex-direction: column;
  transition: background .3s ease-in-out;
}

.list-item:hover {
  background-color: #F7F7F7;
}

.list-item-title {
  word-break: break-all;
  font-weight: 400;
}

.list-item-amount {
  font-weight: 500;
}

.list-item-date {
  font-weight: 300;
  margin: 0;
}

@media (min-width: 40rem) {
  .list-item {
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
  }

  .list-item-amount {
    padding-left: 5rem;
  }

  .list-item-date {
    margin-bottom: 1rem;
  }
}