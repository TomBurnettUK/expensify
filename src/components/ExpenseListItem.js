import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import 'numeral/locales/en-gb';

import './ExpenseListItem.css';

numeral.locale('en-gb');

const ExpenseListItem = props => (
  <Link className="list-item" to={`/edit/${props.id}`}>
    <div>
      <h3 className="list-item-title">{props.description}</h3>
      <p className="list-item-date">
        {moment(props.createdAt).format('MMMM Do, YYYY')}
      </p>
    </div>
    <h3 className="list-item-amount">
      {numeral(props.amount / 100).format('$0,0.00')}
    </h3>
  </Link>
);

export default ExpenseListItem;
