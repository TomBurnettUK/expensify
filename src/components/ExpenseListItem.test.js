import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from './ExpenseListItem';
import expenses from '../fixtures/expenses';

it('should render ExpenseListItem with expense', () => {
  const expense = expenses[0];
  const wrapper = shallow(<ExpenseListItem key={expense.id} {...expense} />);
  expect(wrapper).toMatchSnapshot();
});
