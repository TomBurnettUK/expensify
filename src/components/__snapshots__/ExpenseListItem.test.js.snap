// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/a"
>
  <div>
    <h3
      className="list-item-title"
    >
      Gum
    </h3>
    <p
      className="list-item-date"
    >
      January 1st, 1970
    </p>
  </div>
  <h3
    className="list-item-amount"
  >
    Â£1.95
  </h3>
</Link>
`;
